### DISCUSSION TITLE
How to count words.

### DISCUSSION THREAD

**Discussion by yuuminakamura** [facebook/lexical#1980] [Q&A] [answered]
Hi, I would like to count the content words.
```
       const editorStateTextString = editorState.read(() =>
            $getRoot().getTextContent(),
        );
        console.log('editorStateTextString', editorStateTextString);
        const wordCount = editorStateTextString.split(' ').length -1 ;`
        console.log('wordCount', wordCount);
```
![Screen Shot 2022-04-25 at 12 55 14 PM](https://user-images.githubusercontent.com/46723838/165164789-2303b695-1e4a-4000-8dbf-16d1adfbd150.png)

wordCount is 1 even though there are 3 words. When there is indent, the content will become one word.
in this case editorStateTextString is "applemelonorange"

is there better way I can count the words.

**Comment by trueadm** [comment]
You could try looking at Intl.Segmenter for this. The issue with checking for spaces is that that won't work for many use-cases:

- What about punctuation? e.g. "Hello there!How are you?"
- What about other languages that don't use spaces? i.e. Japanese.

**Comment by zurfyx** [comment]
To expand on @trueadm's answer, the immediate issue you're pointing out is a bug on our side - https://github.com/facebook/lexical/pull/1993

But even after we land the fix keep in mind that nodes can take many shapes. For example, both TextNodes and DecoratorNodes have content (even LineBreakNode). A DecoratorNode may be an image with an a11y label as text. Do you want to count this accessibility label as part of the number of words in your document? This adds up the Intl.Segmenter point for the text content itself.

That said, there's various solutions from less to more flexible:
1. A simple `getTextContent()` or even simpler (`$textContentCurry`)
2. $nodesOfType(TextNode, ..) -> limitation is that it will only track the specific type `text`
3. A `$dfs`, a sorted iteration on which you can choose which nodes to grab its text content from (with `node.getTextContent()`). If you're only interested in TextNodes you can filter by `$isTextNode(node)`

Also, keep in mind that solutions 2. and 3. are more expensive since they have to iterate all over the tree. 1. is O(1) because it's an exception where we build a cache during the reconciliation process as it's a common use case

**Reply by KarthikRaju391** [reply]
Would you mind expanding more on the first approach @zurfyx? And I can't seem to find the $textContentCurry API either in the documentation

**Reply by SalahAdDin** [reply]
> Would you mind expanding more on the first approach @zurfyx? And I can't seem to find the $textContentCurry API either in the documentation

Did you get any solution for it?

**Comment by yuuminakamura** [comment, marked as answer]
@trueadm @zurfyx 

thank you very much for your help.✨
yes I would like to roughly count the words. It would be nice if getTextContent will return the contents which has a space for the indent. I will check the update for the fix https://github.com/facebook/lexical/pull/1993

and will try on option 2 or 3 if needed

**Reply by yuuminakamura** [reply]
Hi @zurfyx 🙂
I am using latest version
and still see the issue with indent. when we indent the words between lines become 1 word using getTextContent()
after https://github.com/facebook/lexical/pull/1993 work
 I am wondering if you guys are going to work on this near the future

**Reply by zurfyx** [reply]
@yuuminakamura - Thanks for the feedback! Sounds like a(nother) bug, let me have a look at this next week (created https://github.com/facebook/lexical/issues/2106)

**Reply by yuuminakamura** [reply]
thank you very much again. I will check https://github.com/facebook/lexical/issues/2106 work
